plugins/modules/mso_schema_site_contract_service_graph_listener.py

```diff
@@ -0,0 +1,744 @@
+        required_if=[
+            ["state", "absent", ["listener"]],
+            ["state", "present", ["listener", "listener_protocol", "listener_port", "rules"]],
+        ],
+    )
+
+    mso = MSOModule(module)
+    site = module.params.get("site")
+
+    # Get site id
+    site_id = mso.lookup_site(site)
+
+    if mso.site_type == "on-premise" or mso.cloud_provider_type != "azure":
+        mso.fail_json(msg="The Site Contract Service Graph Listener is not supported for the site: {0}.".format(site))
+
+    schema = module.params.get("schema")
+    template = module.params.get("template")
+    contract = module.params.get("contract")
+    service_node_index = module.params.get("service_node_index")
+    listener = module.params.get("listener")
+    tenant = module.params.get("tenant")
+    device = module.params.get("device")
+    listener_protocol = module.params.get("listener_protocol")
+    listener_port = module.params.get("listener_port")
+    security_policy = LISTENER_SECURITY_POLICY_MAP.get(module.params.get("security_policy"))
+    ssl_certificates = module.params.get("ssl_certificates")
+    frontend_ip = module.params.get("frontend_ip")
+    rules = module.params.get("rules")
+    state = module.params.get("state")
+
+    if listener_protocol == "https":
+        mso.input_validation("listener_protocol", "https", ["security_policy", "ssl_certificates"], module.params, None, object_name=listener)
+
+    mso_schema = MSOSchema(mso, schema, template)
+    mso_schema.set_template(template)
+    schema_id = mso.lookup_schema(schema, True)
+    mso_schema.set_site(template, site)
+    mso_schema.set_site_contract(contract, False)
+
+    service_graph_ref = mso_schema.schema_objects["site_contract"].details.get("serviceGraphRelationship", {}).get("serviceGraphRef")
+    service_nodes = mso_schema.schema_objects["site_contract"].details.get("serviceGraphRelationship", {}).get("serviceNodesRelationship", [])
+
+    listeners_data = []
+    parent_present = False
+
+    if service_graph_ref is None:
+        mso.fail_json(msg="The site contract: {0} is not associated with a service graph.".format(contract))
+    # Parent object present check
+    else:
+        if service_node_index is not None and service_node_index >= 0:
+            if len(service_nodes) > 0 and len(service_nodes) > service_node_index:
+                service_node = service_nodes[service_node_index]
+                # Query all listeners under a contract service node
+                # The below condition was never false if the service graph was created properly, but condition is required to avoid error
+                if service_node.get("serviceNodeRef") and service_node.get("serviceNodeRef").split("/")[-1] == "node{0}".format(service_node_index + 1):
+                    listeners = service_node.get("deviceConfiguration", {}).get("cloudLoadBalancer", {}).get("listeners", [])
+                    if listeners:
+                        parent_present = True
```

Comment:
Why use this flag and not store the parent object? That case the parent_present flag seems not required.

Reply:
Initially, the site contract contains only serviceNodesRelationship with a list of service node references.

To create a listener, we need to attach the deviceConfiguration.cloudLoadBalancer.listeners with the current parent object. So we need to update the existing object with the PATCH-replace operation. Also it depends on the parent_present flag value.